### ISSUE TITLE
errormessage for getNeighborIdx(idx,j)

### ISSUE THREAD

**Issue by tonimueller** [LeMonADE-project/LeMonADE#21]
Up to now there is no nice errormessage for the function getNeighborIdx(idx, j). I suggest to catch the error from ConnectedDecorator.h and print out the idx and j ?!

``` cpp
try{return vertices.at(idx).getNeighborIdx(j);}
catch(std::runtime_error& exception){
    std::stringstream messagestream;
    messagestream<<"getNeighborIdx(idx, j): j="<<j
                 <<" is out of range for monomer"<<idx<<std::endl;
    throw std::range_error(messagestream.str());
  }
```

**Comment by hrabbel** [comment]
On my machine this costs 10% performance. Suggestion instead:

```
#ifdef DEBUG
    try{
#endif
        return vertices.at(idx).getNeighborIdx(j);
#ifdef DEBUG
    }
    catch(std::runtime_error& exception){
        std::stringstream messagestream;
        messagestream<<"getNeighborIdx(idx, j): j="<<j
                 <<" is out of range for monomer"<<idx<<std::endl;
        throw std::range_error(messagestream.str());
  }
#endif
  }
```

**Comment by tonimueller** [comment]
I hadn't checked the performance...Thus it seems to be the better solution.

**Comment by hrabbel** [comment]
After some discussion about this I have checked the performance again with the two methods above, and also with a third implementation using [] operator instead of at() method. This time I have checked it on a different machine and with two different systems ( see attached files ).
[configs.zip](https://github.com/LeMonADE-project/LeMonADE/files/216997/configs.zip)

The results are roughly the same. Numbers are attempted moves/sec

Dendrimer, save to file every 1e6 mcs
a) current implementation 8.6e6 
b) with try-catch block 7.9e6
c) with [] and no try-catch block 8.2e6

Melt, save to file every 20000 mcs
a) 8.4e6
b) 7.7e6
c) 8.1e6

**Comment by hrabbel** [comment]
Another way with less ifdefs would be

``` cpp
uint32_t   getNeighborIdx(uint32_t idx, uint32_t j) const {
#ifdef DEBUG
    if(j>=getNumLinks(idx)){
        std::stringstream messagestream;
        messagestream<<"getNeighborIdx(idx, j): j="<<j<<" is out of range for monomer"<<idx<<std::endl;
        throw std::range_error(messagestream.str());
    }
#endif /*DEBUG*/
    return vertices.at(idx).getNeighborIdx(j);

}
```

**Comment by MartinWenge** [comment]
We fixed this with #34
